vpr/src/device/rr_graph/rr_graph_obj.h

```diff
@@ -0,0 +1,318 @@
+    bool check_edge_switches() const;
+    bool check_edge_src_nodes() const;
+    bool check_edge_sink_nodes() const;
+    bool check_duplicated_edges() const; /* identify and report any duplicated edges between two nodes */
+    bool check_dangling_nodes() const;   /* identify if there is any dangling nodes in the graph */
+
+    /* Full set checking using listed checking functions*/
+    bool check() const;
+
+  public: //Mutators
+    /* reserve the lists of nodes, edges, switches etc */
+    void reserve_nodes(int num_nodes);
+    void reserve_edges(int num_edges);
+    void reserve_switches(int num_switches);
+    void reserve_segments(int num_segments);
+
+    /* Related to Nodes */
+    RRNodeId create_node(t_rr_type type);
+    RREdgeId create_edge(RRNodeId source, RRNodeId sink, RRSwitchId switch_id);
+    RRSwitchId create_switch(t_rr_switch_inf switch_info);
+    RRSegmentId create_segment(t_segment_inf segment_info);
+
+    void remove_node(RRNodeId node);
+    void remove_edge(RREdgeId edge);
+
+    void set_node_xlow(RRNodeId node, short xlow);
+    void set_node_ylow(RRNodeId node, short ylow);
+    void set_node_xhigh(RRNodeId node, short xhigh);
+    void set_node_yhigh(RRNodeId node, short yhigh);
+    void set_node_bounding_box(RRNodeId node, vtr::Rect<short> bb);
+
+    void set_node_capacity(RRNodeId node, short capacity);
+
+    void set_node_ptc_num(RRNodeId node, short ptc);
+    void set_node_pin_num(RRNodeId node, short pin_id);
+    void set_node_track_num(RRNodeId node, short track_id);
+    void set_node_class_num(RRNodeId node, short class_id);
+
+    void set_node_cost_index(RRNodeId node, short cost_index);
+    void set_node_direction(RRNodeId node, e_direction direction);
+    void set_node_side(RRNodeId node, e_side side);
+    void set_node_R(RRNodeId node, float R);
+    void set_node_C(RRNodeId node, float C);
+    void set_node_segment(RRNodeId node, RRSegmentId segment_index);
+
+    /* Edge related */
+    /* classify the input edges of each node to be configurable (1st part) and non-configurable (2nd part) */
+    void partition_node_in_edges(RRNodeId node);
+    /* classify the output edges of each node to be configurable (1st part) and non-configurable (2nd part) */
+    void partition_node_out_edges(RRNodeId node);
+    void partition_in_edges();  /* classify the input edges of each node to be configurable (1st part) and non-configurable (2nd part) */
+    void partition_out_edges(); /* classify the output edges of each node to be configurable (1st part) and non-configurable (2nd part) */
+    void partition_edges();     /* classify the edges of each node to be configurable (1st part) and non-configurable (2nd part) */
+    void load_switch_C();       /* configure the C of each node with the C of each incoming edge switch */
+
+    void compress();
+    bool validate();
+
+    void clear_nodes();
+    void clear_edges();
+    void clear_switches();
+    void clear_segments();
+    void clear();
```

Comment:
Nodes/edges/switches/segments are inter-related and cross-coupled in the class implementation, so I don't think it makes sense to expose them in the public interface separately. For instance, calling clear_nodes() puts the whole RR graph into an inconsistent state which isn't usable (e.g. while edges could still be iterated through, the edge's node references would be invalid).

I think only the whole class `clear()` method should be exposed publicly. All the other `clear_*()` should be private, since they can't be called safely by clients of this class.

Reply:
These are now private member functions except the clear().